自分の面談日程の表示・登録・編集・削除機能

mashvalue1 opened the pull request:
プルリクエストを作成するにあたって、説明文を書いてみましょう。
各項目に例を載せているので、それに合わせて書いてみて下さい！

やりたかったこと
---
- 自分の面談の一覧表示、新規登録、編集、削除機能の実装

やったこと
----

- interviewモデルの作成、userモデルとのアソシエーションの設定
- interviewsコントローラの作成、必要なactionの実装、ビューの設定

動作確認方法
---

- heroku（https://e-navigator-mashvalue1.herokuapp.com/）へアクセスし、ログインし、index pageへ移動。
- Your interviewsから面談一覧へアクセス（まだ登録されていないため、予定は表示されない）
- Create new interviewへアクセスし、面談を新規登録する。面談一覧へリダイレクトされ、登録された面談が表示されていることを確認する。
- Edit this interviewから面談を編集する。面談一覧にリダイレクトされ、面談情報が更新されていることを確認する。
- Delete this interviewへアクセスし、面談情報が削除されていることを確認する。

その他
---

Commit:
create interview_model and interviews_table

Commit:
create interviews_controller

Commit:
add routing for interviews_controller

Commit:
add user_id-column to interviews_table

Commit:
add association between user and interview

Commit:
create index action / view

Commit:
add : create interview action and some links

面接作成機能の実装、各ページへのリンクを配置

Commit:
add interview edit / update action

Commit:
add interview destroy action

Commit:
add before_aciton : set_user, set_interview

共通アクションをメソッド定義、beforeフィルターで実行

sukechannnn commented:
機能の点でいくつか修正&実装してほしい箇所があるので、コードの修正の前にそちらをお願いします🙏

- `root_path`へのリンクは全てのページに出るようにしてください。ヘッダー部分が良いと思います。
  - もちろん、今後ページが追加されても手で追加するようなことがないようにしてください。
- 面接日程を作成する際に表示される時間が日本時間ではないようです。日本時間にしてください。
- 面接日程が過去の日付けでも作成できてしまうようです。未来だけにしてください。
- ログインや会員登録のところでもそうなのですが、ログインや面接日程の作成に失敗した際にエラーメッセージを表示するようにしてください。

よろしくお願いします！

Commit:
set JST

Commit:
add validation for interview date

Commit:
add flash on successful action

action成功時にフラッシュを表示するように設定

Commit:
add header view

共通するビューをheaderとして部分テンプレートで定義

Commit:
fix date expression

面談一覧の日付フォーマットを設定

Commit:
add interview form view as partial

面談日時のフォームを部分テンプレートとして切り分け

Commit:
fix devise view

メールでのパスワード再設定は利用していないのでコメントアウト

mashvalue1 commented:
以下の修正を行いました。
- 共通で表示するリンクをheaderとして表示
- TimezoneをJSTに変更
- 面談日時のバリデーションを定義
- flash, errorsを表示するように設定

sukechannnn commented on app/models/interview.rb:
```diff
@@ -0,0 +1,8 @@
+  validate :check_datetime
+
+  def check_datetime
+    errors[:base] << "日時は現在以降を設定してください。" if self.interview_date < Time.current
```
いいですね👍
ちなみに、Railsには`past?`というメソッドがあるので、それを使うともっとシンプルに書くことができます。

kielze commented on app/controllers/interviews_controller.rb:
```diff
@@ -0,0 +1,54 @@
+class InterviewsController < ApplicationController
+  before_action :set_user, except: :destroy
```
`set_user`使ってるのいいですね👍

sukechannnn commented on app/controllers/interviews_controller.rb:
```diff
@@ -0,0 +1,54 @@
+  before_action :set_interview, only: [:edit, :update, :destroy]
+
+  def index
+    @interviews = @user.interviews.order("interview_date ASC")
```
ASCの場合は、デフォルトがASCなので指定不要です🙏

sukechannnn replied:
修正したら、コメントに対して下記のようにプルリクエストのハッシュと一緒にコメントしてもらえると、レビュアーが非常に助かります！

sukechannnn replied:
e8508e7 修正を確認しました！

sukechannnn commented on app/controllers/interviews_controller.rb:
```diff
@@ -0,0 +1,54 @@
+class InterviewsController < ApplicationController
+  before_action :set_user, except: :destroy
+  before_action :set_interview, only: [:edit, :update, :destroy]
```
`set_user`と`set_interview `使ってるのいいですね👍

sukechannnn commented on app/views/devise/shared/_links.html.erb:
```diff
@@ -6,9 +6,9 @@
-<% end -%>
+<%#- if devise_mapping.recoverable? && controller_name != 'passwords' && controller_name != 'registrations' %>
+  <%#= link_to "Forgot your password?", new_password_path(resource_name) %><br />
+<%# end -%>
```
ここ不要なら消しちゃっていいですよ！
後で戻すならライブラリを見ればよいので大丈夫です👍

sukechannnn commented on db/schema.rb:
```diff
@@ -10,11 +10,20 @@
+    t.integer "status"
+    t.datetime "created_at", null: false
+    t.datetime "updated_at", null: false
+    t.bigint "user_id"
```
ここは`null: false`にした方がいいですね。でないと、Userに紐付いていない宙ぶらりんなスケジュールが作られてしまう可能性があります。
また、他のカラムに関しても、作成時に入力が必須のものに関しては`null: false`にした方がいいです。

sukechannnn replied:
migation fileで定義している部分なので、修正する際はschema.rbを直接いじらないように気をつけてください🙏

sukechannnn commented on app/views/layouts/_header.html.erb:
```diff
@@ -0,0 +1,18 @@
+          <%= link_to "Sign Out", destroy_user_session_path, method: :delete %>
+        </li>
+      <% end %>
+  </ul>
```
今回はデザインは気にしないで良いと言った手前アレなのですが、headerがlistで表示されているのは流石に微妙な気がするので、せめてtableで横に並ぶようにしてもらえると良いかなと思います🙏

sukechannnn commented:
コミットメッセージに何をやったか説明をきちんと書いているの、とても素晴らしいです👏
是非継続してください！

Commit:
fix @interviews var

ASCの記述を削除

Commit:
delete comment

Commit:
fix validation

Commit:
add constraints for interviews table

以下の制約を追加
interview_date 	not_null
user_id 			not_null, foreign_key: true
status 			not_null, default: 0

Commit:
add enum for status

面談の承認状態に対するenumを設定

Commit:
install bottstrap-sass

Commit:
update navbar with bootstrap

bootstrapを用いてヘッダを記述

Commit:
fix include order and update error message and flash

includeの順序を変更、エラーメッセージとフラッシュにcssを設定

Commit:
update user edit form using bootstrap

プロフィールの編集フォームをbootstrapで設定

Commit:
update interview form using bootstrap

Commit:
update interview list using bootstrap

mashvalue1 commented:
遅くなりましたが修正しました。
- past?メソッドの利用
- ASCの記述を削除
- コメントアウトを削除
- interviewsテーブルのカラムに対する制約の設定
- レイアウトの設定（bootstrapの導入）

sukechannnn commented:
良さそうでした！
１章お疲れ様です！
LGTM👍

<img src="https://img.gifmagazine.net/gifmagazine/images/2550/original.gif">
